Count only mapped fields towards docvalue_fields limit

cbuescher opened the pull request:
Currently we count every field requested in the search request bodies
'docvalue_fields' section towards the limit defined by
the 'max_docvalue_fields_search' index setting which defaults to 100. This can
be a problem e.g. if the user searches across several indices with some fields
present in one index but not the other and has to add the joint set of field
names to the query. We currently trip the limit even if the number of actually
mapped fields in each index is below the limit.
This change adds a step to distiguish between mappend and unmapped fields and
only count the former towards the limit.

Closes #63730

Commit:
Count only mapped fields towards docvalue_fields limit

Currently we count every field requested in the search request bodies
'docvalue_fields' section towards the limit defined by
the 'max_docvalue_fields_search' index setting which defaults to 100. This can
be a problem e.g. if the user searches across several indices with some fields
present in one index but not the other and has to add the joint set of field
names to the query. We currently trip the limit even if the number of actually
mapped fields in each index is below the limit.
This change adds a step to distiguish between mappend and unmapped fields and
only count the former towards the limit.

Closes #63730

elasticmachine commented:
Pinging @elastic/es-search (:Search/Mapping)

Commit:
fix yaml

Commit:
Skip yaml test for mixed cluster since message changed

cbuescher commented on rest-api-spec/src/main/resources/rest-api-spec/test/search/30_limits.yml:
```diff
@@ -85,18 +85,22 @@ setup:
 
+  - skip:
+      version: " - 7.99.99"
+      reason: waiting for backport
```
I dislike having to do this here just because I changed the error message slightly. Happy do get some better ideas what we do in such cases in other places if possible.

jtibshirani replied:
For me the old message was just as clear, and lets us avoid having to update the test. But if we want to keep it, I wonder if we could just remove this test since it's already covered by `SearchServiceTests#testMaxDocvalueFieldsSearch`?

jtibshirani commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchDocValuesContext.java:
```diff
@@ -33,20 +32,25 @@
-    public static FetchDocValuesContext create(Function<String, Set<String>> simpleMatchToFullName,
-                                               int maxAllowedDocvalueFields,
-                                               List<FieldAndFormat> fieldPatterns) {
+    public static FetchDocValuesContext create(MapperService mapperService, List<FieldAndFormat> fieldPatterns) {
```
I think this signature change undoes some of @javanna's work in https://github.com/elastic/elasticsearch/pull/63239. The easiest fix would be to keep the function parameter, and add a new one `QueryShardContext::isFieldMapped`.

As a side note, if we end up passing a lot of functions to avoid exposing `MapperService`, I wonder if we should pull out a lightweight abstraction similar to `FieldTypeLookup` and pass it around instead?

javanna replied:
thanks for catching this Julie. I am still addressing the original problem: we carry MapperService around too often when it's not needed. I do see the increasing need for functions which is definitely a signal that we dont have the right abstractions. I am using functions in some places but I see it more as a temporary work-around. I do think once we remove getMapperService from QueryShardContext we should reconsider the functions needed. A field type lookup abstraction may also be needed as I would like to split DocumentMapper into some sub-components that can be shared independently. To be seen very soon :)

cbuescher replied:
Instead of using MapperService, I pass in QueryShardContext now which has methods to resolve field name patterns, allows looking up the setting and look up whether a field is mapped. I can go back to passing in three parameters (functions and the setting itself) if you think this is not ideal for any reason.

jtibshirani commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchDocValuesContext.java:
```diff
@@ -33,20 +32,25 @@
             for (String fieldName: fieldNames) {
                 fields.add(new FieldAndFormat(fieldName, field.format));
+                // only count the mapped fields towards the limit fedinde by IndexSettings.MAX_DOCVALUE_FIELDS_SEARCH_SETTING
+                if (mapperService.fieldType(fieldName) != null) {
```
For efficiency could we only create a new `FieldAndFormat` if the field is mapped? I think we'll end up dropping the field anyways, not even having an entry in the response for it.

Also small typo, fedinde -> defined.

cbuescher replied:
You are right, I only see the "fields" really being used in "FetchDocValuePhase#getProcessor" and we seem to disregard fields without a mapping there. I will update the comments on the class though to reflect this for future usage.

jtibshirani commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchDocValuesContext.java:
```diff
@@ -33,20 +32,25 @@
             throw new IllegalArgumentException(
-                "Trying to retrieve too many docvalue_fields. Must be less than or equal to: [" + maxAllowedDocvalueFields
-                    + "] but was [" + fields.size() + "]. This limit can be set by changing the ["
+                "Trying to retrieve too many mapped docvalue_fields. Must be less than or equal to: [" + maxAllowedDocvalueFields
```
For me the old message was just as clear, and lets us avoid having to update the REST tests.

jtibshirani reviewed (commented):
Thanks @cbuescher for such a speedy fix !

Commit:
Merge branch 'master' into fix-63730

Commit:
iter

cbuescher commented:
@jtibshirani thanks for the update, I changed the signature and simplified as you requested. I also removed the current static factory methods in favour of only having a simple constructor. I think in this case its sufficient. Let me know what you think.

jtibshirani commented on rest-api-spec/src/main/resources/rest-api-spec/test/search/30_limits.yml:
```diff
@@ -85,6 +85,10 @@ setup:
 ---
 "Docvalues_fields size limit":
 
+  - skip:
```
Do we want to remove this skip now?

cbuescher replied:
sure, missed that

Commit:
remove skip

cbuescher commented:
@elasticmachine  run elasticsearch-ci/1

cbuescher commented:
@jtibshirani thanks for the last review, does this look okay now?

jtibshirani reviewed (approved):
Looks good to me.

cbuescher commented:
@jtibshirani thanks, will merge
